src/deploy/functions/prepare.ts

```diff
@@ -4,19 +4,48 @@ import { Options } from "../../options";
 import { ensureCloudBuildEnabled } from "./ensureCloudBuildEnabled";
 import { functionMatchesAnyGroup, getFilterGroups } from "./functionsDeployHelper";
 import { logBullet } from "../../utils";
-import { getFunctionsConfig, getEnvs, prepareFunctionsUpload } from "./prepareFunctionsUpload";
+import { getFunctionsConfig, prepareFunctionsUpload } from "./prepareFunctionsUpload";
+import { previews } from "../../previews";
 import { promptForFailurePolicies, promptForMinInstances } from "./prompts";
 import * as args from "./args";
 import * as backend from "./backend";
 import * as ensureApiEnabled from "../../ensureApiEnabled";
-import { FirebaseError } from "../../error";
 import * as functionsConfig from "../../functionsConfig";
+import * as functionsEnv from "../../functions/env";
 import { needProjectId } from "../../projectUtils";
 import * as runtimes from "./runtimes";
 import * as validate from "./validate";
 import * as utils from "../../utils";
 import { logger } from "../../logger";
 
+function getEnvs(options: {
+  firebaseConfig: { [key: string]: any };
+  functionsSource: string;
+  projectId: string;
+  projectAlias?: string;
+}): Record<string, string> {
+  const { firebaseConfig, functionsSource, projectId, projectAlias } = options;
+
+  let envs = {
+    FIREBASE_CONFIG: JSON.stringify(firebaseConfig),
+    GCLOUD_PROJECT: projectId,
+  };
+  if (previews.dotenv) {
+    envs = {
+      ...envs,
```

Comment:
Since we never want to allow a customer to override this, even though the logic of `functionsEnv.load` won't allow these to be set I'd still prefer to put `...envs` last.